49-jackal/technical_specification.md

```diff
@@ -0,0 +1,208 @@
+- To be implemented into the schemapack library with a dedicated CLI command
+- The mermaid language is used as output (visualization must be performed
+  elsewhere)
+- The following example schemapack:
+  ```yaml
+  schemapack: 0.2.0
+  description: A simple schemapack # with no root class but with descriptions.
+  classes:
+    File:
+      id:
+        propertyName: alias
+      content: ../../content_schemas/File.schema.json
+    Dataset:
+      id:
+        propertyName: alias
+      content: ../../content_schemas/Dataset.schema.json
+      relations:
+        files:
+          targetClass: File
+          mandatory:
+            origin: true
+            target: true
+          multiple:
+            origin: true
+            target: true
+    ```
+- Should be transpiled to mermaid as:
+  ```
+  erDiagram
+      File }|--|{ Dataset: "Dataset.files"
+      Dataset {
+          string dac_contact
+      }
+      File {
+          string filename
+          string format
+          string checksum
+          int size
+      }
+  ```
+- visualized as (e.g. view this on GitHub for the diagram to be automatically rendered):
+  ```mermaid
+  erDiagram
+      File }|--|{ Dataset: "Dataset.files"
+      Dataset {
+          string dac_contact
+      }
+      File {
+          string filename
+          string format
+          string checksum
+          int size
+      }
+  ```
+- In agreement with the schemapack concept, there
+  is a clear separation between content properties and
+  relations:
+  - content properties are shown as attributes of the
+    entities
+  - relations are shown as arrows connecting entities
+- the name of the arrows are composed of the name of
+  the class defining the relation and the name of the
+  relation property; this makes the directionality of
+  the relation clear
+- crow foot notation is used to adequatly model both
+  the modality (`mandatory` specs in schemapack) and
+  the cardinality (`multiple` specs in schemapack),
+  please see https://vertabelo.com/blog/crow-s-foot-notation/
+  for further details
+- Since schemapack makes no assumption about content
+  schemas of individual classes (except that they
+  discribe JSON object), not every content schema can
+  be fully visualized. Thus only the properties of
+  the top-level object of the content schema should be
+  visualized. Thereby, non-primitive types (including
+  properties that define nested structures such as
+  further objects or arrays) are annotated with the type
+  "Custom".
+- There should also be an option to skip content
+  information from the diagram, like that:
+  ```
+  erDiagram
+      File }|--|{ Dataset: "Dataset.files"
+      Dataset {}
+      File {}
+  ```
+- Which is visualizd as:
+  ```mermaid
+  erDiagram
+      File }|--|{ Dataset: "Dataset.files"
+      Dataset {}
+      File {}
+  ```
+
+### III. Autogenerate documentation from schemapack specs
```

Comment:
I'm not sure if this should be part of this epic or if we should rather plan our documentation strategy first. This is nice to have seeing schemapack as a standalone solution for a general problem, but probably not needed in this form in the GHGA context.

Reply:
Removed